File: Lista de Exercícios/Python/Banco.py
# Questão 3 e 15
class ContaBancaria:
    def __init__(self, titular, saldo_inicial=0):
        self.__titular = titular
        self.__saldo = saldo_inicial

    @property
    def titular(self):
        return self.__titular

    @property
    def saldo(self):
        return self.__saldo

    @saldo.setter
    def saldo(self, valor):
        raise AttributeError("Não é possível modificar o saldo diretamente. Use os métodos depositar() ou sacar().")

    def depositar(self, valor):
        if valor > 0:
            self.__saldo += valor
            return True
        return False

    def sacar(self, valor):
        if valor > 0:
            if self.__saldo >= valor:
                self.__saldo -= valor
                return True
            else:
                raise SaldoInsuficienteException(self.__saldo, valor)
        return False

class SaldoInsuficienteException(Exception):
    def __init__(self, saldo_atual, valor_saque):
        self.saldo_atual = saldo_atual
        self.valor_saque = valor_saque
        super().__init__(f"Saldo insuficiente. Saldo atual: R${saldo_atual:.2f}, Valor do saque: R${valor_saque:.2f}")

conta_maria = ContaBancaria("Maria", 1000)

print(f"Titular: {conta_maria.titular}")
print(f"Saldo inicial: R${conta_maria.saldo:.2f}")

if conta_maria.depositar(500):
    print(f"Depósito realizado. Novo saldo: R${conta_maria.saldo:.2f}")
else:
    print("Falha no depósito.")

if conta_maria.sacar(200):
    print(f"Saque realizado. Novo saldo: R${conta_maria.saldo:.2f}")
else:
    print("Falha no saque.")

try:
    if conta_maria.sacar(2000):
        print(f"Saque realizado. Novo saldo: R${conta_maria.saldo:.2f}")
except SaldoInsuficienteException as e:
    print(f"Falha no saque: {e}")

try:
    conta_maria.saldo = 5000
except AttributeError as e:
    print(f"Erro ao tentar modificar o saldo: {e}")

File: Lista de Exercícios/Java/Banco.java
package Lista de Exercícios.Java;

// Questão 3 e 15
class ContaBancaria {
    private String titular;
    private double saldo;

    public ContaBancaria(String titular, double saldoInicial) {
        this.titular = titular;
        this.saldo = saldoInicial;
    }

    public String getTitular() {
        return titular;
    }

    public double getSaldo() {
        return saldo;
    }

    public void setSaldo(double valor) {
        throw new IllegalArgumentException("Não é possível modificar o saldo diretamente. Use os métodos depositar() ou sacar().");
    }

    public boolean depositar(double valor) {
        if (valor > 0) {
            saldo += valor;
            return true;
        }
        return false;
    }

    public boolean sacar(double valor) throws SaldoInsuficienteException {
        if (valor > 0) {
            if (saldo >= valor) {
                saldo -= valor;
                return true;
            } else {
                throw new SaldoInsuficienteException(saldo, valor);
            }
        }
        return false;
    }
}

class SaldoInsuficienteException extends Exception {
    public SaldoInsuficienteException(double saldoAtual, double valorSaque) {
        super(String.format("Saldo insuficiente. Saldo atual: R$%.2f, Valor do saque: R$%.2f", saldoAtual, valorSaque));
    }
}

public class Banco {
    public static void main(String[] args) {
        ContaBancaria contaMaria = new ContaBancaria("Maria", 1000);

        System.out.printf("Titular: %s%n", contaMaria.getTitular());
        System.out.printf("Saldo inicial: R$%.2f%n", contaMaria.getSaldo());

        if (contaMaria.depositar(500)) {
            System.out.printf("Depósito realizado. Novo saldo: R$%.2f%n", contaMaria.getSaldo());
        } else {
            System.out.println("Falha no depósito.");
        }

        try {
            if (contaMaria.sacar(200)) {
                System.out.printf("Saque realizado. Novo saldo: R$%.2f%n", contaMaria.getSaldo());
            }
        } catch (SaldoInsuficienteException e) {
            System.out.printf("Falha no saque: %s%n", e.getMessage());
        }

        try {
            contaMaria.sacar(2000);
        } catch (SaldoInsuficienteException e) {
            System.out.printf("Falha no saque: %s%n", e.getMessage());
        }

        try {
            contaMaria.setSaldo(5000);
        } catch (IllegalArgumentException e) {
            System.out.printf("Erro ao tentar modificar o saldo: %s%n", e.getMessage());
        }
    }
}
